crates/iceberg/src/spec/partition.rs

```diff
@@ -235,6 +261,179 @@ impl UnboundPartitionSpec {
     }
 }
 
+/// Trait for common functions between [`PartitionSpec`], [`UnboundPartitionSpec`] and [`PreservedPartitionSpec`]
+pub trait UnboundPartitionSpecInterface<T: PartitionFieldInterface> {
```

Comment:
Hi, compared to having a trait, I personally feel that building an `enum` for this would be better. Maybe we could have:

```rust
enum PartitionSpec {
  Bound(BoundPartitionSpec),
  Unbound(UnboundPartitionSpec),
}
```

Users can convert between bound and unbound versions based on their needs. This makes it easier to find the differences between various partition specs.

I believe this can avoid some complexity around the trait system and narrow down our API surface. What do you think?

Reply:
Sounds like a good plan - and thanks for taking the time to review this @Xuanwo!

I will give it a shot and see how it feels.